src/main/jbake/content/blog/2016/10/14/hawkular-apm-polyglot-zipkin.adoc

```diff
@@ -0,0 +1,190 @@
+using one line docker-compose command. Overall there are seven containers including Apache Kafka 
+for collecting span data. The complete source code with instructions how to run it can be found on
+https://github.com/hawkular/hawkular-apm/tree/master/examples/polyglot-zipkin[Github].
+
+== Instrumentation Libraries
+There are several http://zipkin.io/pages/existing_instrumentations.html[instrumentation libraries]. 
+Developers have to choice one for targeted 
+language and framework they are using. If there is no instrumentation for a specific
+framework they have to report spans to the server manually, thus having an instrumentation
+library for a popular web framework is a big advantage (not limited to web).
+ 
+In a typical microservice environment it is also needed to add instrumentation for outgoing 
+requests so one application uses two instrumentations. One for incoming requests and one 
+for outgoing. These two instrumentation have to share context in order to accurately set id of 
+a parent span.
+
+Our polyglot applications are instrumented with following instrumentations:
+
+ * Java (JAX-RS): https://github.com/openzipkin/brave[brave]
+ * Java Dropwizard (JAX-RS): https://github.com/smoketurner/dropwizard-zipkin[dropwizard-zipkin] uses brave
+ * Apache Cassandra: https://github.com/thelastpickle/cassandra-zipkin-tracing[cassandra-zipkin-tracing] uses brave
+ * JavaScript (express, cujoJS): https://github.com/openzipkin/zipkin-js[zipkin-js]
+ * Ruby (rack): https://github.com/openzipkin/zipkin-ruby[zipkin-ruby]
+ * Python (Pyramid): https://github.com/Yelp/pyramid_zipkin[pyramid_zipkin]
+ * Python (Bravado): https://github.com/Yelp/swagger_zipkin[swagger_zipkin]
+
+These libraries are typically developed by a community and therefore reported data may differ. For 
+example a key of binary annotation representing HTTP URL can be `http.url` or `http.uri` or even 
+`http.path`. Some of the libraries do not report HTTP status code. Name of a span should represent
+HTTP method but some instrumentations put there URL or even other data (cassandra-zipkin-tracing).
+
+== Run the Polyglot Application
+Follow the instructions on 
+https://github.com/hawkular/hawkular-apm/tree/master/examples/polyglot-zipkin[Github] 
+to run the environment and start Hawkular APM server.
+
+Execute following requests which we will analyze in the following sections.
+-----
+1. curl -ivX GET 'http://localhost:3001/nodejs/hello'
+2. curl -ivX GET 'http://localhost:3000/dropwizard/A
+3. curl -ivX GET 'http://localhost:3000/dropwizard/asyncTwoOutgoingCalls'
+4. curl -ivX POST -H 'Content-Type: application/json' 'http://localhost:3001/nodejs/createUser' -d '{"name": "jdoe"}'
+-----
+
+== Hawkular APM
+In Hawkular APM user interface we are going to look at distributed tracing tab. If you run previous 
+commands there should be some data for analysis. 
+
+Following figure show services interactions for selected initial endpoint (fourth request). 
```

Comment:
Need to make a deliberate mistake occasionally to make sure people are paying attention 😀